Abstract evaluation of switch cases (simple).

neilmacintosh opened the pull request:
Adds abstract evaluation for switch cases. Effectively treats the switch case as a series of if-else statements to achieve the same outcome.

Currently does not support case blocks that result in an abrupt completion through `return/throw/continue`. That is work-to-go in the future.

A future PR will also expand on the tests to provide more coverage (and look for corner cases).

Test plan: Added more tests to serializer/abstract/Switch.js

hermanventer commented on src/evaluators/SwitchStatement.js:
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
+    }
+
+    // TODO work out what to do with Reference
+    if (result instanceof Reference) {
```
Can't happen.

neilmacintosh replied:
Yay! Will remove it.

hermanventer commented on src/evaluators/SwitchStatement.js:
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
+          throw ReportNoPossiblyNormalCompletionError();
+        }
+
+        result = GetValue(r, result, realm);
```
Just use UpdateEmpty instead of GetValue.

neilmacintosh replied:
ok.

neilmacintosh commented on src/evaluators/SwitchStatement.js:
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
+
+    let selector = CaseSelectorEvaluation(test, strictCode, env, realm);
+    let selectionResult = computeBinary(realm, "===", input, selector);
+    if (selectionResult instanceof ConcreteValue) {
```
Oh nice, I didn't realize!

neilmacintosh commented on src/evaluators/SwitchStatement.js:
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
+        let node = c.consequent[i];
+        let r = env.evaluateCompletion(node, strictCode);
+
+        // TODO understand Reference and how to deal with it better
```
Yay. That helps shed light on what Reference is. ;-)

neilmacintosh commented on src/evaluators/SwitchStatement.js:
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
+      }
+    }
+
+    invariant(selectionResult instanceof AbstractValue);
```
yep!

neilmacintosh commented on src/evaluators/SwitchStatement.js:
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
+        if (r instanceof Reference) {
+          throw ReportNoReferenceError();
+        } else if (r instanceof PossiblyNormalCompletion) {
+          throw ReportNoPossiblyNormalCompletionError();
```
Of course. Note that in this case I am not making up exception classes. I am calling functions with those names that return the usual FatalError. I just wanted to wrap them up rather than repeat the code to report the diagnostic and throw the error each time I wanted to do it. So is using a wrapper function like this to do what you described a problem?

hermanventer replied:
It is easy to misread this pattern, which is what happened to me. The other reason you don't want to do this is that you end up with very different looking code when you decide to actually recover from an error.

neilmacintosh replied:
Good point! Especially in Javascript where things are much more free flowing than my most recent implementation language (not good or bad...just different ;-)).

neilmacintosh commented on src/evaluators/SwitchStatement.js:
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
+          throw ReportNoPossiblyNormalCompletionError();
+        }
+
+        result = GetValue(r, result, realm);
```
Makes sense, but Isn't that what happens? I think GetValue() makes result an AbruptCompletion in that case and then the test in the loop header will break out of the loop.

hermanventer replied:
Missed that. Putting the check in the loop header makes that logic non local and hard to spot when reading the code incrementally.

neilmacintosh replied:
Yep. I was trying to avoid having two if-then-break tests scattered around the loop, but...ugly as it is....it is probably much easier to understand. I'll change it.

neilmacintosh commented on src/evaluators/SwitchStatement.js:
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
+    if (result instanceof BreakCompletion) {
+      return result.value;
+    } else if (result instanceof AbruptCompletion) {
+      throw ReportNoAbruptCompletionError();
```
Makes sense!

neilmacintosh replied:
Interestingly, this then caused my tests to fail. It looks as though I have a little bit of work to do around handling Abrupt and PossiblyNormalCompletions correctly (beyond completions arising from break statements). So I've left it to error out of abstract evaluation if it sees a case that ends up in AbruptCompletion for now (just without the use of the helper function to do so). That way I can focus on finishing up this PR and addressing the completion issues (and i/o effects, expanding tests etc) in the next one.

neilmacintosh commented on src/evaluators/SwitchStatement.js:
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
+      return result.value;
+    } else if (result instanceof AbruptCompletion) {
+      throw ReportNoAbruptCompletionError();
+    } else if (result instanceof Completion) {
```
excellent. I can simplify this then.

neilmacintosh commented:
I think I've addressed all the comments except the one about throwing AbruptCompletions. That exposed a bug in my handling of them which I'd like to postpone to the next PR.

hermanventer commented on src/evaluators/SwitchStatement.js:
```diff
@@ -90,32 +56,32 @@ function AbstractCaseBlockEvaluation(
-        } else if (r instanceof PossiblyNormalCompletion) {
-          throw ReportNoPossiblyNormalCompletionError();
+        if (r instanceof PossiblyNormalCompletion) {
+          let msg = "This operation is not yet supported on case blocks that result in a PossiblyNormalCompletion.";
```
PossiblyNormalCompletion is an implementation detail that should not appear in a error message that is read by a developer using Prepack.

neilmacintosh replied:
ok. makes sense. I think your next comment has the punchline here ;-)

hermanventer commented on src/evaluators/SwitchStatement.js:
```diff
@@ -90,32 +56,32 @@ function AbstractCaseBlockEvaluation(
-          throw ReportNoPossiblyNormalCompletionError();
+        if (r instanceof PossiblyNormalCompletion) {
+          let msg = "This operation is not yet supported on case blocks that result in a PossiblyNormalCompletion.";
+          realm.reportIntrospectionError(msg);
```
You want to create a CompilerDiagnostic, for which you'll need a new error code, which implies that you'll update the Wiki with a nice explanation of how such a situation arises and how to avoid it.

neilmacintosh replied:
Aha. I see now. I'll go look into how to do that and update this PR appropriately.

neilmacintosh replied:
Ok. Can you confirm that the correct place to describe the (hopefully temporary) limitation is here: https://prepack.io/frequently-asked-questions.html?

neilmacintosh replied:
Ignore that, I found it!

neilmacintosh replied:
Ok. Created the page and uploading another diff to see if I have the error handling right.

hermanventer commented on src/evaluators/SwitchStatement.js:
```diff
@@ -90,32 +56,32 @@ function AbstractCaseBlockEvaluation(
-    } else if (result instanceof Completion) {
-      return result.value;
+      // TODO correct handling of PossiblyNormal and AbruptCompletion
+      let msg = "This operation is not yet supported on case blocks that throw.";
```
Or return, or continue.

neilmacintosh replied:
Yep.

hermanventer commented on test/serializer/abstract/Switch.js:
```diff
@@ -61,12 +61,11 @@ switch (x) {
-  case 2: z11 = 13; break;
-  default: throw 2; break;
-}
+// switch (x) {
```
Should this be commented out?

neilmacintosh replied:
Ah! No. Nice catch.

hermanventer commented on src/evaluators/SwitchStatement.js:
```diff
@@ -90,32 +56,32 @@ function AbstractCaseBlockEvaluation(
       for (let i = 0; i < c.consequent.length && !(result instanceof Completion); ++i) {
```
!(result instanceof Completion) is always true here.

neilmacintosh replied:
Ah yes...leftover from moving out the breaks. I'll kill it off, thanks.

hermanventer commented:
Time to update the PR title and summary.

hermanventer commented on src/evaluators/SwitchStatement.js:
```diff
@@ -36,16 +39,133 @@ function CaseSelectorEvaluation(
+      return AbstractCaseEvaluation(caseIndex + 1);
+    } else if (caseIndex >= cases.length) {
+      // this is the stop condition for our recursive search for a matching case.
+      // we tried every available case index and if nothing matches we return
```
and if -> and since

neilmacintosh replied:
thanks!

neilmacintosh commented:
Ok. Updated title, summary and tag. Added proper emission of a compiler error (at least I think so) and a wiki page for the new error.
